File: typescript/strain/strain.ts
export function keep<T>(collection: T[], predicate: (element: T)  => boolean ): T[] {
    // no op
    const kept: T[] = []
    collection.forEach((element) => {
        if (predicate(element)) {
            kept.push(element)
        }
    })
    return kept
}

export function discard<T>(collection: T[], predicate: (element: T)  => boolean ): T[] {
    // no op
    const discarded: T[] = []
    collection.forEach((element) => {
        if (!predicate(element)) {
            discarded.push(element)
        }
    })
    return discarded

}

File: java/strain/src/main/java/Strain.java
import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class Strain {

  public static <T> List<T> keep(List<T> list, Predicate<T> p) {
    List<T> results = new ArrayList<>();

    for (T item : list) {
      if (p.test(item)) {
        results.add(item);
      }
    }

    return results;
  }

  public static <T> List<T> discard(List<T> list, Predicate<T> p) {
    List<T> results = new ArrayList<>();

    for (T item : list) {
      if (!p.test(item)) {
        results.add(item);
      }
    }

    return results;
  }
}
